src/find/matchers/perm.rs

```diff
@@ -0,0 +1,480 @@
+                        self.state = ParserState::GatheringPermissions;
+                    } else {
+                        return self.error();
+                    }
+                }
+                'r' => {
+                    if let ParserState::GatheringPermissions = self.state {
+                        self.bit_pattern |= self.category_bit_pattern << 2;
+                    } else {
+                        return self.error();
+                    }
+                }
+                'w' => {
+                    if let ParserState::GatheringPermissions = self.state {
+                        self.bit_pattern |= self.category_bit_pattern << 1;
+                    } else {
+                        return self.error();
+                    }
+                }
+                'x' => {
+                    if let ParserState::GatheringPermissions = self.state {
+                        self.bit_pattern |= self.category_bit_pattern;
+                    } else {
+                        return self.error();
+                    }
+                }
+                't' => {
+                    if let ParserState::GatheringPermissions = self.state {
+                        self.bit_pattern |= 0o1000;
+                    } else {
+                        return self.error();
+                    }
+                }
+                's' => {
+                    if let ParserState::GatheringPermissions = self.state {
+                        // if we're setting group bits, then set the set-group-id bit
+                        if (self.category_bit_pattern & 0o010) == 0o010 {
+                            self.bit_pattern |= 0o2000;
+                        }
+                        // if we're setting user bits, then set the set-user-id bit
+                        if (self.category_bit_pattern & 0o100) == 0o100 {
+                            self.bit_pattern |= 0o4000;
+                        }
+                    } else {
+                        return self.error();
+                    }
+                }
+                ',' => {
+                    if let ParserState::GatheringPermissions = self.state {
+                        self.state = ParserState::GatheringCategories;
+                        self.category_bit_pattern = 0;
+                    } else {
+                        return self.error();
+                    }
+                }
+                _ => {
+                    return self.error();
+                }
+            };
+            Ok(())
+        }
+    }
+
+    pub fn parse(string_value: &str) -> Result<(u32, ComparisonType), Box<Error>> {
+        // safe to unwrap as the regex is a compile-time constant.
+        let re = Regex::new("^([/-]?)([0-7]+)$").unwrap();
+
+        // have we been given a simple octal based string (e.g. /222)?
+        if let Some(m) = re.captures(string_value) {
+            // all these unwraps are safe because we checked the string in the regex above
```

Comment:
Technically, this is not safe (in the sense of panic being impossible) because of overflow.